### ISSUE TITLE
Feature Request: chan.select()

### ISSUE THREAD

**Issue by eugeneware** [brentropy/chan#7]
Hi Brent,

Really loving this module. It's great to be able to use CSP-style concurrency in node.js!

The only roadblock I've found when using `chan` is that there's no easy way to be able to do the equivalent of a select.

Ideally I'd love to be able to write the following co code:

``` js
var co = require('co'),
    chan = require('chan');
co(function *(){
  var ch1 = chan(),
      ch2 = chan();

  var ch = yield chan.select([ch1, ch2]);
  switch (ch) {
    case ch1:
      var val1 = yield ch1;
      break;

    case ch2:
      var val2 = yield ch2;
      break;
  }
})();
```

There are lots of times where being able to use multiple channels (such as one for a timeout for example) is really useful, and currently there is no easy way to atomically wait for activity on multiple channels.

**Comment by brentropy** [comment]
I am glad you like `chan`. I can see how the select functionality could be useful. I will look in to adding something like this. Thanks for the suggestion Eugene.

**Comment by eugeneware** [comment]
Hi Brent, I actually had another idea about how to achieve the same result, as it's more a synchronization issue of multiple yieldables in co, so I wrote a simple module called [`co-select`](https://github.com/eugeneware/co-select) that can do the job:

``` js
var co = require('co'),
    select = require('co-select'),
    chan = require('chan');
co(function *(){
  var ch1 = chan(),
      ch2 = chan();

  var first = yield select([ch1, ch2]);
  switch (first.caller) {
    case ch1:
      var val1 = first.value;
      break;

    case ch2:
      var val2 = first.value;
      break;
  }
})();
```

If you just want the first value from a channel, you can pass `true` through as the last argument:

``` js
var co = require('co'),
    select = require('co-select'),
    chan = require('chan');
co(function *(){
  var ch1 = chan(),
      ch2 = chan();

  var firstValue = yield select([ch1, ch2], true);
})();
```

**Comment by eugeneware** [comment]
Oops. Spoke too soon. `co-select` won't actually solve the issue, because if a value from `ch1` is read, then we lose the future value from `ch2`.

So, probably needs to be implemented in the `chan` library where you have knowledge of the channel state and don't have to actually read from all the channels in advance to determine which one is ready to be read.

**Comment by eugeneware** [comment]
I had a crack at adding `select()` to `chan` in #8

**Comment by brentropy** [comment]
Resolved by pull request #9. Merged.
